tests/test_bms_enforce_root.py

```diff
@@ -0,0 +1,49 @@
+from typing import Callable, Optional
+
+import numpy as np
+import pytest  # noqa: 401
+
+from autora.skl.bms import BMSRegressor
+
+
+def generate_x_2d(start=-1, stop=1, num=40):
+    step = abs(stop - start) / num
+    x2 = np.mgrid[start:stop:step, start:stop:step].reshape(2, -1).T
+    return x2
+
+
+def transform_through_primitive_softmax_2d(x: np.ndarray) -> np.ndarray:
+    return np.array([y for y in [softmax_2d(x, y) for x, y in x]])
+
+
+def run_test_primitive_fitting_2d(
+    X: np.ndarray,
+    transformer: Callable,
+    verbose: bool = False,
+    custom_primitive: Optional[Callable] = None,
+    root=None,
+):
+    y = transformer(X)
+    regressor = BMSRegressor(epochs=30)
+    regressor.fit(X, y.ravel(), custom_ops=[custom_primitive], root=root)
+    # root is none or is in custom ops
+    assert root is None or root in regressor.custom_ops.values()
+    # root is none or it is the root of the fitted model
+    assert root is None or root is regressor.custom_ops[regressor.model_.root.value]
```

Comment:
This is much better! I was wondering if we could also test that the fixed root is part of the final model returned by BMS? E.g., if we specify a softmax as a custom primitive and fix that as a root, then shouldn't the final BMS tree contain that primitive? We could include an assert that checks for that...

Reply:
I believe that is what the second assert statement does actually. Notice the mention of regressor.model_ in that line.

Reply:
The reason it mentions custom_ops is because trees store all of their operations as strings rather than callable functions. ```custom_ops``` is a dictionary that takes the string and returns the callable function.

Reply:
Got it, thanks for the clarification!